### ISSUE TITLE
docker swarm replicas end up using the same volume when #replicas > #available workers

### ISSUE THREAD

**Issue by TheCount** [docker/docker#26445]
---------------------------------------------------
BUG REPORT INFORMATION
---------------------------------------------------

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 16
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 43
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local flocker
 Network: bridge host null overlay
Swarm: active
 NodeID: 1jkithwhj6tvt1ef8eq4ts5wp
 Is Manager: true
 ClusterID: 8gkvszkwb73drb07l7lo5hy3k
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.77.103
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-31-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 3.859 GiB
Name: manager3
ID: J5WM:PDDN:WZQT:CJDN:EH5E:HYTO:5USM:24WM:KMAR:UFXM:W62I:FEKX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Labels:
 provider=generic
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
3 virtualboxen for playing around, all of them both swarm managers and workers, set up with vagrant and docker machine generic.

**Steps to reproduce the issue:**
1. ````docker service create --replicas 4 --mount type=volume,src=test,dst=/data --name test busybox /bin/sh -c "while true; do sleep 1; date >> /data/Test.txt; done"````

**Describe the results you received:**

The service having more replicas than workers currently available, one node ended up with two replicas of the service, and they are both using the same volume, e. g.

```
# tail Test.txt
Fri Sep  9 15:59:04 UTC 2016
Fri Sep  9 15:59:04 UTC 2016
Fri Sep  9 15:59:05 UTC 2016
Fri Sep  9 15:59:05 UTC 2016
Fri Sep  9 15:59:06 UTC 2016
Fri Sep  9 15:59:06 UTC 2016
Fri Sep  9 15:59:07 UTC 2016
Fri Sep  9 15:59:07 UTC 2016
Fri Sep  9 15:59:08 UTC 2016
Fri Sep  9 15:59:08 UTC 2016
```

On the other node with just one replica, the results are as expected.

**Describe the results you expected:**

Each replica gets its own dedicated volume, e. g.

```
# tail Test.txt
[…]
Fri Sep  9 15:59:04 UTC 2016
Fri Sep  9 15:59:05 UTC 2016
Fri Sep  9 15:59:06 UTC 2016
Fri Sep  9 15:59:07 UTC 2016
Fri Sep  9 15:59:08 UTC 2016
```

**Additional information you deem important (e.g. issue happens only occasionally):**

I haven't tested this with globally scoped volumes. Obviously, two replicas using the same volume could lead to data corruption.

--------

So, I'm very new to docker swarm. Looking around, this seems to be related to #25446, i. e. no scalable stateful docker services possible. So, what is the plan how this might work in the future then? In my (possibly naïve) imagination it would be something like this:

1. A docker swarm service consists (among other things) of n replicas.
1. A replica is a container and a volume set associated with the service (for stateless services, the volume set is empty, so the replica is essentially just a container in this case).
1. Multiple volume sets can be associated with a service. 
1. A volume set can be in the "attached" or in the "unattached" state.
1. If a replica is created and the host it is created on has an unattached volume associated with the service available, it is attached to that replica (with globally scoped volumes, either all hosts have a set available or all hosts don't). Otherwise, a new volume set is created.
1. If a container dies, its attached volume set becomes unattached.
1. No volume set can be attached to two replicas at the same time.

That scheme would make both updates work well and inhibit possible data corruption by two containers writing to the same data.

**Comment by cpuguy83** [comment]
This is the expected behavior.
Swarm supports mounts, ie mounting in a volume.
Swarm does not currently orchestrate volumes. The syntax is very purposefully `--mount` and not `--volume` for this reason.

The idea for orchestrated volumes would probably be task-scoped volumes.
A task is higher level than a container but lower-level than a service.
This means that, for instance, `docker swarm create -v /foo --replicas 2` might actually have two volumes that it manages, one for each task it creates.

There's a lot to figure out here, though. Since tasks can move around we need to be careful about orchestrating both the data itself (ie, mobility) as well as the task (data gravity).

**Comment by TheCount** [comment]
OK, I now understand that this is the currently expected behaviour given the very new state of development. It's just not the expected/desired behaviour.

On the other hand, does docker really have to orchestrate the data (i. e., the persistent state) and/or "task" itself, or should this rather be left to the service? As I understand it, two of the most important differences going from container to service are

1. The lifetime of a container is bounded in that on the next planned maintenance or unplanned outage, it will be replaced/gone. The lifetime of the service knows no such bounds.
1. In a container, the volume is the data. In a service, volumes are "merely" an optimisation feature required for practical purposes, and docker does not need to make too many guarantees about them.

To illustrate my points, consider the case of a postgresql database. Furthermore, let us consider three idealised types of events:

1. (Planned1) Service update,
1. (Planned2) Docker swarm host update, including host reboot,
1. (Unplanned) Boom. One docker swarm host totally kaput, for whatever reason.

I'll assume that (Planned1) happens much more frequently than (Planned2), which in turn happens much more frequently than (Unplanned).

If you were to run your database in a single container without a data volume, even (Planned1) would destroy all your data. Therefore, a data volume is essential for operation here, to survive (Planned1) and (Planned2). Of course, you can't survive (Unplanned) in this setup, which is why we want services instead of containers in the first place (manually, or hopefully in the near future with docker swarm on a large scale).

As part of a service, your postgres containers work a little differently. When such a container is brought up, it will first talk to the nearest consul (or other dkv service), check the state of affairs and, depending on the outcome, do a pg_basebackup/wal sync with the current database master, initialise an empty cluster, and/or take part in a master/slave election, or more complicated things. In any case, it isn't docker's concern. The curious thing, though, is that *in principle* such a container would not need a data volume at all if the service has a sufficient number of replicas. The service can survive all three of the above events just by replicating the entire database into all of its containers every time. From a practical perspective, this would be, of course, a nightmare. Hence, data volumes.

So, to make such a service work in practice, what guarantees would docker have to give to the service? Only the following rather weak one:

"Look, service, if a fresh replica of yours is started, I'll try, on a best effort basis, to give its container a data volume previously used by another replica of yours that has since died, but be prepared that the volume might be just empty."

Let's see what happens in each of the three events defined above with such a policy in place.

(Planned1): docker halts the container, unattaches the volume, replaces the container with one based on the updated image, reattaches the same volume, starts the container. Inside the container, at best, nothing more than would happen with a postgres restart happens (if the upgrade was snappy enough), at worst a failover was executed. Not much data has to be moved, negligible churn.

(Planned2): docker shuts down on the host and unattaches the volume. A new replica is started on another host. If the volume is globally scoped, the unattached volume can be reattached as in (Planned1), negligible churn. If the volume is locally scoped, docker gives the new replica either an old volume of this service that happens to be lying around on that host, or a fresh empty one. In any case, some churn is expected here, but that's what you get for having locally scoped volumes.

(Unplanned): The volume is either lost (local scope) or left in a possibly unsafe state (global scope). A new replica is started on another host, proceeding as in (Planned2). Churn in all cases, but fortunately, the number of unplanned outages per replica is small.

So, long story short, there doesn't seem to be that much to figure out on docker's part since most of the onus of stateful service coordination is on the service anyway. Am I missing something?

**Comment by rdehouss** [comment]
I was seeing this "problem" in a different way, but probably I don't have enough experience to have the correct idea.

I thought that we could consider a container as stateless by default. It would become stateful thanks to a volume.
So, if a volume is already "mounted", it cannot be mounted a second time and another one (with same name) is created and mounted.
The app in the container would detect that it's a new "volume" (empty folder for example) and would initialize it like it would at first launch.

So, at the end, a container is "stupid" until it gets sth to become "smart" (data in the volume).

The only problem I see with this approach is if the container "identify itself" via an IP or a task id and stores it in the volume, then, this id could be different depending on the volume attached.

To be clearer:

- Container A.1 starts on Node 101, it gets the volume "volumeA.someRandomNumber" on Node 101
- Container A.2 starts on Node 104, it gets the volume "volumeA.someOtherRandomNumber" on Node 104

Now, Container A.1 needs to move on Node 103, it gets a new volume "volumeA.someOtherOtherRandomNumber" on Node 103 and so, needs to re-initialize, but it's "not an issue" because it's like "a new container/instance", "like we scaled up".

Then, another Container A.3 starts (scale up) on Node 101, it gets the (pre-existing) volume "volumeA.someRandomNumber" on Node 101 and so, "it looks like" Container A.3 is Container A.1 but who cares because the containers are stateless without volume, so, no "id" should be assigned from external. Optionaly, the containers could generate an "ID" (random) and store it in the volume and so, they could be "named", it's just that they could vanish (if they move node) or new ones could be created.

That schema also work with shared storage for the volumes, it's just that there would be multiple "volumeA.XXX" next to each other and the mount would be "random" and each container would become the one defined by their volume.
That's the idea of distributed computing, we don't care who the jobs are until they define it themselves.
If they need somehow a "fixed" storage, it should be a separated DB.

What am I missing for this to work?
